Simplified/AxisService/NYPLAxisDRMAuthorizer.swift

```diff
@@ -0,0 +1,81 @@
+//
+//  NYPLAxisDRMAuthorizer.swift
+//  Simplified
+//
+//  Created by Raman Singh on 2021-04-13.
+//  Copyright © 2021 NYPL Labs. All rights reserved.
+//
+
+import Foundation
+
+#if AXIS
+@objcMembers
+class NYPLAxisDRMAuthorizer: NSObject, NYPLDRMAuthorizing {
+  
+  static let deviceIDKey = "NYPLAxisDRMAuthorizerKey"
+  
+  static let sharedInstance = NYPLAxisDRMAuthorizer()
+  
+  static var deviceID: String {
+    if
+      let id = UserDefaults.standard.string(forKey: NYPLAxisDRMAuthorizer.deviceIDKey) {
+      return id
+    }
+    
+    let id = UUID().uuidString
+    UserDefaults.standard.set(id, forKey: NYPLAxisDRMAuthorizer.deviceIDKey)
+    return id
+  }
+  
+  /// TODO: To be fixed in SIMPLY-3692. 
+  var workflowsInProgress: Bool {
+    return false
+  }
+  
+  /// We're always returning true since Axis does not deauthorize a user.
+  func isUserAuthorized(_ userID: String!,
+                        withDevice device: String!) -> Bool {
+   return true
+  }
+  
+  /// TODO: SIMPLY-3692 will have code that fixes the issue.
+  /// As of now, we're creating a deviceID which stays saved until the user uninstalls the app.
+  func authorize(withVendorID vendorID: String!,
+                 username: String!,
+                 password: String!,
+                 completion: ((Bool, Error?, String?, String?) -> Void)!) {
+    
+    /*
+     In NYPLSignInBusinessLogic+OE::drmAuthorize:username:password:loggingContext,
+     we expect this method to take some time. After calling this method, we
```

Comment:
since this implementation of `authorize(withVendorID:username:password:completion:)` does not need any delay and its execution is synchronous (we just need to call `completion(true, nil, NYPLAxisDRMAuthorizer.deviceID, username)`, correct?) I think we should just avoid the horrible hack we have in `NYPLSignInBusinessLogic+DRM::drmAuthorize(username:password:loggingContext:)` entirely. Can we refactor that code so that for OE we don't need to run any of that code? There's no need to comply to that craziness now that we don't need Adobe anymore. 

One quick way could be to move that `drmAuthorize(username:password:loggingContext:)` function out into an extension, and compile that only for SimplyE, and then write another extension for OE. That's just the first I could think of though.

Reply:
Agreed. That makes a lot more sense. However, the only thing we're doing different from Adobe is calling the selector after delay and cancelling the selector in completion handler. I was of the opinion that the following code should do the job.

```
private func drmAuthorize(username: String,
                            password: String?,
                            loggingContext: [String: Any]) {

    let vendor = userAccount.licensor?["vendor"] as? String

    Log.info(#file, """
      ***DRM Auth/Activation Attempt***
      Token username: \(username)
      Token password: \(password ?? "N/A")
      VendorID: \(vendor ?? "N/A")
      """)

    drmAuthorizer?
      .authorize(withVendorID: vendor,
                 username: username,
                 password: password) { success, error, deviceID, userID in
                  #if FEATURE_DRM_CONNECTOR
                  // make sure to cancel the previously scheduled selector
                  // from the same thread it was scheduled on
                  NYPLMainThreadRun.asyncIfNeeded { [weak self] in
                    if let self = self {
                      NSObject.cancelPreviousPerformRequests(withTarget: self)
                    }
                  }
                  #endif

                  Log.info(#file, """
                    Activation success: \(success)
                    Error: \(error?.localizedDescription ?? "N/A")
                    DeviceID: \(deviceID ?? "N/A")
                    UserID: \(userID ?? "N/A")
                    ***DRM Auth/Activation completion***
                    """)

                  var success = success

                  if success, let userID = userID, let deviceID = deviceID {
                    NYPLMainThreadRun.asyncIfNeeded {
                      self.userAccount.setUserID(userID)
                      self.userAccount.setDeviceID(deviceID)
                    }
                  } else {
                    success = false
                    NYPLErrorLogger.logLocalAuthFailed(error: error as NSError?,
                                                       library: self.libraryAccount,
                                                       metadata: loggingContext)
                  }

                  self.finalizeSignIn(forDRMAuthorization: success,
                                      error: error as NSError?)
    }

    #if FEATURE_DRM_CONNECTOR
    NYPLMainThreadRun.asyncIfNeeded { [weak self] in
      self?.perform(#selector(self?.dismissAfterUnexpectedDRMDelay), with: self, afterDelay: 25)
    }
    #endif
  }
```

Reply:
However this does make the method a bit harder to understand and modify. Your approach is better and I think we should go with that.

Reply:
yeah, if I understand correctly it doesn't seem like we need that complex implementation of `drmAuthorize(username:password:loggingContext:)` on Open eBooks.